scout/adapter/mongo/matchmaker.py

```diff
@@ -131,15 +131,16 @@ def case_mme_delete(self, case_obj, user_obj):
 
         """
         institute_obj = self.institute(case_obj["owner"])
-        # create events for subjects removal from Matchmaker this cas
+        # create events for subjects removal from Matchmaker this case
+        link = f"/{institute_obj['_id']}/{case_obj['display_name']}"
```

Comment:
and here, right?

Reply:
Same as above

Reply:
url_for becomes a problem when you try for instance to delete a user using the command line, since you are not in the app context:

```
(py38) chiararasi@ChiaraRMBP:~/Documents/work/GITs/scout$ scout --demo delete user -m clark.kent@mail.com
2022-06-21 08:27:55 ChiaraRMBP.local scout.commands.base[18659] INFO Running scout version 4.52
2022-06-21 08:27:55 ChiaraRMBP.local scout.commands.base[18659] DEBUG Debug logging enabled.
2022-06-21 08:27:55 ChiaraRMBP.local scout.adapter.client[18659] INFO Connected to MongoDB 3.6.18
2022-06-21 08:27:55 ChiaraRMBP.local scout.adapter.mongo.base[18659] DEBUG Database name: scout-demo
2022-06-21 08:27:55 ChiaraRMBP.local scout.server.app[18659] INFO Matchmaker Exchange extension enabled
2022-06-21 08:27:55 ChiaraRMBP.local scout.utils.scout_requests[18659] DEBUG Sending GET request to http://localhost:9020/nodes
2022-06-21 08:27:55 ChiaraRMBP.local urllib3.connectionpool[18659] DEBUG Starting new HTTP connection (1): localhost:9020
2022-06-21 08:27:55 ChiaraRMBP.local urllib3.connectionpool[18659] DEBUG http://localhost:9020 "GET /nodes HTTP/1.1" 200 3
2022-06-21 08:27:55 ChiaraRMBP.local scout.commands.delete.delete_command[18659] INFO Running scout delete user
2022-06-21 08:27:55 ChiaraRMBP.local scout.adapter.mongo.case[18659] INFO Fetching case internal_id
User can't be removed because MatchMaker Exchange submissions (n. cases=1) are associated to it. Reassign patients to another user? [y/N]: y
Assign patients to user with email: wonderwoman@mail.com
2022-06-21 08:28:03 ChiaraRMBP.local scout.adapter.mongo.case[18659] INFO Fetching case internal_id
2022-06-21 08:28:03 ChiaraRMBP.local scout.adapter.mongo.case[18659] INFO Updating case internal_id
2022-06-21 08:28:03 ChiaraRMBP.local scout.adapter.mongo.case[18659] DEBUG Removing the following unused keys from updated case: dict_keys(['cnv_report', 'coverage_qc_report', 'gene_fusion_report_research', 'multiqc'])
2022-06-21 08:28:03 ChiaraRMBP.local scout.adapter.mongo.case[18659] INFO Case updated
2022-06-21 08:28:03 ChiaraRMBP.local scout.commands.delete.delete_command[18659] ERROR Application was not able to create a URL adapter for request independent URL generation. You might be able to fix this by setting the SERVER_NAME config variable.
(py38) chiararasi@ChiaraRMBP:~/Documents/work/GITs/scout$
```

Reply:
Yes, routing is not up if the app is not. There would be workarounds I guess, but they would likely be worse than just a string. We in general have problems with the lack of a user for the cli commands to generate good audit tracking for cli actions, but that is another issue.

Buuut, I thought delete user was in `@with_appcontext`? I wonder if there is not something else here - maybe its in the extension as well?

Reply:
I think @with_appcontext is not sufficient, I'm reading that you also need a request context in order to generate urls in that way. Or you need to provide a SERVER_NAME in the config file:

https://code.luasoftware.com/tutorials/flask/things-you-should-know-about-flask-server-name/

String looks easier in this case

Reply:
Hm, that sounds weird; appcontext should be enough for routing. request context should not be needed. I more get the feeling that the one of the click.prompt or confirm calls drops the app context before you get to the mme_reassign call? Perhaps an explicit `with app.app_context():` for it? But otherwise, naturally its fine with a string. The thing is though, if it has trouble with `url_for()`, it will probably have trouble with getting app config as well, right?

Reply:
I think it's because of an absence of request whenever you run the commands: Perhaps this tutorial explains it better:  https://developpaper.com/configuration-tutorial-of-server_name-domain-name-item-in-pythons-flask-framework/

Reply:
Also true that in the tutorial they are generating external urls..

Reply:
Perhaps it is solvable by explicitly using the app request context when creating the url:

with app.app_context():
    url = url_for('page', param="hello")

But seems more complicated than just composing the string. I'll just leave it like this in this file. I'm using url_for in the beacon code since it's always invoked from the web app.

Reply:
Yes, thats absolutely fine! Said so somewhere, but probably mixed the functions up.

I agree, the tutorial is for external links. It makes some convoluted sense a request context is needed to determine how to make the url for an external link; it would try to figure out what server name the referral was to, since it doesn't really know its "name" unless its given statically with SERVER_NAME, and that is only found in the request. But for storing in the db we dont want an external url, only the routing internal part, and for that app context should be sufficient.

I predict we would run into trouble if trying to actually use the app config as well while in that click interim context or prompt and so on. `with app.app_contex()` before the call to mme_reassign.